lib/winapi/filemanip.c

```diff
@@ -241,6 +246,146 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+        NULL);
+    if (sourceHandle == INVALID_HANDLE_VALUE) {
+        return FALSE;
+    }
+
+    status = NtQueryInformationFile(
+        sourceHandle,
+        &ioStatusBlock,
+        &networkOpenInformation,
+        sizeof(networkOpenInformation),
+        FileNetworkOpenInformation);
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    RtlInitAnsiString(&targetPath, lpNewFileName);
+    InitializeObjectAttributes(&objectAttributes, &targetPath, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtCreateFile(
+            &targetHandle,
+            FILE_GENERIC_WRITE,
+            &objectAttributes,
+            &ioStatusBlock,
+            &networkOpenInformation.AllocationSize,
+            networkOpenInformation.FileAttributes,
+            0,
+            bFailIfExists ? FILE_CREATE : FILE_SUPERSEDE,
+            FILE_SYNCHRONOUS_IO_NONALERT | FILE_NON_DIRECTORY_FILE | FILE_SEQUENTIAL_ONLY);
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    status = NtAllocateVirtualMemory(&readBuffer,
+                                      0,
+                                      &readBufferRegionSize,
+                                      MEM_RESERVE | MEM_COMMIT,
+                                      PAGE_READWRITE);
+    if (!NT_SUCCESS(status)) {
+        NtClose(sourceHandle);
+        NtClose(targetHandle);
+        DWORD lastError = RtlNtStatusToDosError(status);
+        SetLastError(lastError);
+        return FALSE;
+    }
+
+    while (TRUE) {
+        const BYTE *bufferPos = readBuffer;
+        if (!ReadFile(sourceHandle, readBuffer, readBufferRegionSize, &bytesRead, NULL)) {
+            NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+            DeleteHandle(targetHandle);
+            NtClose(sourceHandle);
+            NtClose(targetHandle);
+            return FALSE;
+        }
+
+        if (!bytesRead) {
+            break;
+        }
+
+        while (bytesRead > 0) {
+            DWORD bytesWritten = 0;
+            if (!WriteFile(targetHandle, bufferPos, bytesRead, &bytesWritten, NULL)) {
+                NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+                DeleteHandle(targetHandle);
+                NtClose(sourceHandle);
+                NtClose(targetHandle);
+                return FALSE;
+            }
+            bytesRead -= bytesWritten;
+            bufferPos += bytesWritten;
+        }
+    }
+
+    status = NtFreeVirtualMemory(&readBuffer, &readBufferRegionSize, MEM_RELEASE);
+    if (!NT_SUCCESS(status)) {
+      DWORD lastError = RtlNtStatusToDosError(status);
+      SetLastError(lastError);
+    }
```

Comment:
Should the method fail if it fails to free the temp memory? Technically I think the copy could still succeed, and I don't know that there's anything the caller could do about the leak other than abort entirely.